app/scripts/modules/core/src/instance/Instances.tsx

```diff
@@ -1,6 +1,133 @@
+import * as React from 'react';
+import { UIRouterContext } from '@uirouter/react-hybrid';
+import { Transition } from '@uirouter/core';
+import autoBindMethods from 'class-autobind-decorator';
+import { Subscription } from 'rxjs';
+
 import { IInstance } from 'core/domain';
 
-export interface InstancesProps {
+const equalForKeys = (keys: string[], left: any, right: any): boolean =>
+    keys.reduce((acc, key) => acc && left[key] === right[key], true);
+
+export interface IInstancesProps {
   instances: IInstance[];
   highlight?: string;
 }
+
+export interface IInstancesState {
+  detailsInstanceId: string;
+}
+
+@UIRouterContext
+@autoBindMethods
+export class Instances extends React.Component<IInstancesProps, IInstancesState> {
+  // context from enclosing UIView
+  public static contextTypes = {
+    router: React.PropTypes.object,
+    parentUIViewAddress: React.PropTypes.object,
+  };
+
+  private subscription: Subscription;
+
+  constructor(props: IInstancesProps) {
+    super(props);
+    this.state = {
+      detailsInstanceId: null,
+    }
+  }
+
+  public render() {
+    const instances = (this.props.instances || []).sort((a, b) => a.launchTime - b.launchTime);
+
+    return (
+      <div className="instances">
+        {instances.map(instance => (
+          <Instance
+            key={instance.id}
+            instance={instance}
+            active={this.state.detailsInstanceId === instance.id}
+            highlight={this.props.highlight}
+            onInstanceClicked={this.handleInstanceClicked}
+          />
+        ))}
+      </div>
+    )
+  }
+
+  public componentDidMount() {
+    this.subscription = this.context.router.globals.success$.subscribe(this.onStateChange);
+  }
+
+  private onStateChange(transition: Transition) {
+    const isShowingDetails = !!/\.instanceDetails/.exec(transition.to().name);
+    const detailsInstanceId = isShowingDetails ? transition.params().instanceId : null;
+    this.setState({ detailsInstanceId });
+  }
+
+  public componentWillUnmount() {
+    this.subscription && this.subscription.unsubscribe();
+  }
+
+  public shouldComponentUpdate(nextProps: IInstancesProps, nextState: IInstancesState) {
+    const propsKeys: [keyof IInstancesProps] = ['instances', 'highlight'];
+    if (!equalForKeys(propsKeys, this.props, nextProps)) {
+      return true;
+    }
+
+    if (this.state.detailsInstanceId !== nextState.detailsInstanceId) {
+      const ids = nextProps.instances.map(x => x.id);
+      if (ids.includes(this.state.detailsInstanceId) || ids.includes(nextState.detailsInstanceId)) {
```

Comment:
This could be optimized to just iterate once instead of twice or thrice. Instead of iterating to create a map, then iterating twice to check if the map includes something:

```
nextProps.instances.some(x => x.id === this.state.detailsInstanceId || x.id === nextState.detailsInstanceId)
```